lib/services/elasticsearch.js

```diff
@@ -153,665 +173,682 @@ class ElasticSearch extends Service {
+  {
+    const esRequest = {
+      index: this._getESIndex(index, collection),
+      body: { query },
+      scroll: '30s',
+      from,
+      size,
+      refresh
+    };
 
-    assertWellFormedRefresh(esRequestBulk);
+    if (! _.isPlainObject(query)) {
+      return this.reject('empty_query');
+    }
 
-    esRequestSearch.body = request.input.body;
-    esRequestSearch.scroll = '30s';
-    esRequestBulk.body = [];
+    return this.client.deleteByQuery(esRequest)
+      .then(({ body }) => {
+        const apiResponse = {
+          total: body.total,
+          deleted: body.deleted,
+          failures:
+            body.failures.map(({ shardId, reason }) => {
+              return {
+                shardId,
+                reason
+              };
+            })
+        };
 
-    if (!esRequestSearch.body.query ||
-      !(esRequestSearch.body.query instanceof Object)) {
-      return Bluebird.reject(this.getError('empty_query'));
-    }
+        return apiResponse;
+      })
+      .catch(error => this.esWrapper.reject(error));
+  }
 
-    return getAllIdsFromQuery(this.client, addActiveFilter(esRequestSearch))
-      .then(ids => {
-        ids.forEach(id => {
-          esRequestBulk.body.push(
-            {update:
-              {_index: esRequestBulk.index,
-                _type: esRequestBulk.type,
-                _id: id}
-            });
-          esRequestBulk.body.push(
-            {doc:
-              {_kuzzle_info:
-                { active: false,
-                  deletedAt: ts,
-                  updater: this._getUserId(request)}}
-            });
-        });
+  /**
+   * Create an empty collection. Mapping will be applied if supplied.
+   *
+   * @param {string} index - Index name
+   * @param {string} collection - Collection name
+   * @param {object} mappings - Collection mappings in ES format
+   *
+   * @returns {Promise} { }
+   */
+  createCollection (index, collection, mappings = {}) {
+    const esRequest = {
+      index: this._getESIndex(index, collection),
+      body: _.merge(mappings, this.config.commonMapping)
+    };
 
-        if (esRequestBulk.body.length === 0) {
-          return Bluebird.resolve({ids: []});
+    return this.kuzzle.indexCache.exists(index)
+      .then(indexExists => {
+        if (! indexExists) {
+          this.throw('index_does_not_exist');
         }
 
-        return this.client.bulk(esRequestBulk)
-          .then(() => this.refreshIndexIfNeeded(esRequestBulk, {ids}))
-          .catch(error => Bluebird.reject(
-            this.esWrapper.formatESError(error))
-          );
+        return this.kuzzle.indexCache.exists(index, collection);
+      })
+      .then(exists => {
+        if (exists) {
+          return this.updateMapping(index, collection, mappings);
+        }
+
+        this._checkMappings(esRequest.body.mappings);
+
+        return this.client.indices.create(esRequest)
+          .then(() => {
+            const apiResponse = {};
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Immediately return this expression instead of assigning it to the temporary variable "apiResponse". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1488)